lib/mocha.js

```diff
@@ -247,6 +258,28 @@ Mocha.prototype._growl = function(runner, reporter) {
 };
 
 /**
+ * Get require.cache files (node only)
+ *
+ * @api private
+ */
+Mocha.prototype._getRequireCache = function() {
+  return Object.keys(require.cache);
+};
+
+/**
+ * Flush test files from require.cache
+ *
+ * @api private
+ */
+Mocha.prototype._flushRequireCache = function() {
+  this._getRequireCache().forEach(function(filePath) {
+    if (this._requireCache.indexOf(filePath) === -1) {
```

Comment:
Thank you for such a full explanation, I finally got your point.
Just added dropRequireCacheBeforeRun() method for this and updated tests.